foundation-models/openai/src/main/java/com/sap/ai/sdk/foundationmodels/openai/StreamConverter.java

```diff
@@ -0,0 +1,108 @@
+package com.sap.ai.sdk.foundationmodels.openai;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Spliterator.NONNULL;
+import static java.util.Spliterator.ORDERED;
+
+import io.vavr.control.Try;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Spliterators;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.hc.core5.http.HttpEntity;
+
+@Slf4j
+class StreamConverter {
+
+  /** see {@link BufferedReader#DEFAULT_CHAR_BUFFER_SIZE} */
+  static final int BUFFER_SIZE = 8192;
+
+  @FunctionalInterface
+  private interface ReadHandler<T> {
+    /**
+     * Read next entry for Stream.
+     *
+     * @return The next entry, or {@code null} when no further entry can be read.
+     * @throws IOException if no entry can be read anymore, unexpected.
+     */
+    @Nullable
+    T readEntry() throws IOException;
+  }
+
+  @FunctionalInterface
+  private interface CloseHandler {
+    /** Close handler to be called when Stream terminated. */
+    void close();
+  }
+
+  @RequiredArgsConstructor
+  private static class HandledIterator<T> implements Iterator<T> {
+    private final ReadHandler<T> readHandler;
+    private final CloseHandler stopHandler;
+    private boolean done = false;
+    private T next = null;
+
+    @Override
+    public boolean hasNext() {
```

Comment:
A note somewhere here that states parallel execution is NOT supported (because we don't see a usecase for it) might be useful.

Reply:
While it's not typical, we have no control or guarantee that parallel execution won't happen. But since there is no use case for it we should make the decision to not be thread safe explicit 😉

Reply:
Added JavaDoc to emphasize on the fact 👍